plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/SizeBasedBlockingQ.java

```diff
@@ -0,0 +1,190 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Queue implementation to accept events based on their storage attribute. If size of queue is breached, then transfer
+ * event is rejected.
+ */
+public class SizeBasedBlockingQ extends AbstractLifecycleComponent {
+    private static final Logger log = LogManager.getLogger(SizeBasedBlockingQ.class);
+
+    protected final LinkedBlockingQueue<Item> queue;
+    protected final Lock lock;
+    protected final Condition notEmpty;
+
+    protected final AtomicLong currentSize;
+    protected final ByteSizeValue capacity;
+    protected final AtomicBoolean closed;
+    protected final ExecutorService executorService;
+    protected final int consumers;
+
+    /**
+     * Constructor to create sized based blocking queue.
+     */
+    public SizeBasedBlockingQ(ByteSizeValue capacity, ExecutorService executorService, int consumers) {
+        this.queue = new LinkedBlockingQueue<>();
+        this.lock = new ReentrantLock();
+        this.notEmpty = lock.newCondition();
+        this.currentSize = new AtomicLong();
+        this.capacity = capacity;
+        this.closed = new AtomicBoolean();
+        this.executorService = executorService;
+        this.consumers = consumers;
+    }
+
+    @Override
+    protected void doStart() {
+        for (int worker = 0; worker < consumers; worker++) {
+            Thread consumer = new Consumer(queue, currentSize, lock, notEmpty, closed);
+            executorService.submit(consumer);
+        }
+    }
+
+    /**
+     * Add an item to the queue
+     */
+    public void produce(Item item) throws InterruptedException {
+        if (item == null || item.size <= 0) {
+            throw new IllegalStateException("Invalid item input to produce.");
+        }
+
+        if (currentSize.get() + item.size >= capacity.getBytes()) {
+            throw new S3TransferRejectedException("S3 Transfer queue capacity reached");
+        }
+
+        final Lock lock = this.lock;
+        final AtomicLong currentSize = this.currentSize;
+        lock.lock();
+        try {
+            if (currentSize.get() + item.size >= capacity.getBytes()) {
+                throw new S3TransferRejectedException("S3 Transfer queue capacity reached");
+            }
+            if (closed.get()) {
+                throw new AlreadyClosedException("Transfer queue is already closed.");
+            }
+            queue.put(item);
+            currentSize.addAndGet(item.size);
+            notEmpty.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public int getSize() {
+        return queue.size();
+    }
+
+    public boolean canProduce(long contentLength) {
+        return (currentSize.get() + contentLength) < capacity.getBytes();
```

Comment:
We never want to produce any item if it is exceeding the available capacity of the queue. That is a hard line we have set. Just contentLength will not honor the current utilization of the queue. canProduce as the name also says makes sure it does. Now coming to the comment and this use case, if current availability cannot fit the item in the queue then it can anyways be retried by caller but the point I was trying to make is that even if queue is completely free then a file larger than its capacity can never be accepted and the only option is to serialize the entire upload.

Reply:
Yes it should be. I was also accounting for the current utilization but then the effect would be that we will start redirecting all other uploads to slow client when queue is full instead of rejecting them. I will make the change.